packages/gatsby/src/internal-plugins/prod-404/gatsby-node.js

```diff
@@ -4,18 +4,24 @@ const { actions } = require(`../../redux/actions`)
 const PROD_404_PAGE_PATH = `/404.html`
 
 let page404 = null
-exports.onCreatePage = ({ page, store, actions }) => {
+emitter.on(`CREATE_PAGE`, action => {
   // Copy /404/ to /404.html as many static site hosts expect
   // site 404 pages to be named this.
   // https://www.gatsbyjs.org/docs/how-to/adding-common-features/add-404-page/
-  if (!page404 && /^\/?404\/?$/.test(page.path)) {
-    actions.createPage({
-      ...page,
-      path: PROD_404_PAGE_PATH,
-    })
-    page404 = page
+  if (!page404 && /^\/?404\/?$/.test(action.payload.path)) {
+    store.dispatch(
+      actions.createPage(
+        {
+          path: PROD_404_PAGE_PATH,
+          component: action.payload.component,
+          context: action.payload.context,
+        },
+        action.plugin
+      )
+    )
+    page404 = true
```

Comment:
`page404` shoudl be page object and not boolean - few lines below we try to get `page404.path` to see wether we need to delete `404.html` when original 404 page is deleted

Reply:
adjusted in https://github.com/gatsbyjs/gatsby/pull/31236/commits/aa78b624c601960b855e505a15c19d16ce7348c0 (i opened separate PR against this one to work on it separately without messing your branch)